src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+import { DidDocument } from '@decentralized-identity/did-common-typescript';
+import * as Base58 from 'bs58';
+import { Cas } from './Cas';
+import Multihash from './Multihash';
+import { WriteOperation, OperationType } from './Operation';
 import Transaction from './Transaction';
-import { WriteOperation } from './Operation';
 
 /**
- * Interface for caching all the processed DIDs and related states.
- * This interface is mainly useful for creating a mock instances for testing purposes.
+ * VersionId identifies the version of a DID document. We use the hash of the
+ * operation that produces a particular version of a DID document as its versionId.
+ * This usage is guaranteed to produce unique VersionId's since the operation contains
+ * as one of its properties the previous VersionId. Since the operation hash is
+ * just a string we alias VersionId to string.
+ *
+ * With this usage, the operation hash serves two roles (1) an identifier for an operation
+ * (2) an identifier for the DID document produced by the operation. In the code below,
+ * we always use VersionId in places where we mean (2) and an OperationHash defined below
+ * when we mean (1).
  */
-export interface DidCache {
-  /** The transaction that was COMPLETELY processed. */
-  readonly lastProcessedTransaction?: Transaction;
+export type VersionId = string;
+
+/**
+ * Alias OperationHash to string - see comment above
+ */
+export type OperationHash = string;
+
+/**
+ * Function type that updates a Did document given an operation. This would be instantiated
+ * with a function that implements json patch application. Using the interface instead of
+ * the actual function hides details of json patching from Did cache.
+ */
+export interface DidDocumentUpdater {
+  (didDoc: DidDocument, operation: WriteOperation): DidDocument;
+}
+
+/**
+ * Function type that creates a Did document given a create/recover operation.
+ */
+export interface DidDocumentCreator {
```

Comment:
Design:
1. I feel it is an overkill to define a DidDcoumentCreator interface because there is really only one implementation and we ought to test that implementation whenever we can, and since this implementation does not involve network calls, I don't see huge benefit for mocking and injecting it, but see the downside of more complex code (well, not crazy complex, but still).
2. Similar sentiment for DidDocumentUpdator, the desire to not hide patching details from the DidCache is valid, but I see it can be done by just declaring the function `static`.

Reply:
This is mostly to get a working DidCache without needing to have a fully implemented json patch functions. Once we have those we could simply substitute those implementations. 

I have taken your suggestion and removed these from being exported to users of the class (e.g., the constructor now no longer needs these functions).